core/trino-main/src/test/java/io/trino/execution/scheduler/TestFaultTolerantStageScheduler.java

```diff
@@ -0,0 +1,564 @@
+    private static final PlanFragmentId SOURCE_FRAGMENT_ID_1 = new PlanFragmentId("1");
+    private static final PlanFragmentId SOURCE_FRAGMENT_ID_2 = new PlanFragmentId("2");
+    private static final PlanNodeId TABLE_SCAN_NODE_ID = new PlanNodeId("table_scan_id");
+
+    private static final CatalogName CATALOG = new CatalogName("catalog");
+
+    private static final InternalNode NODE_1 = new InternalNode("node-1", URI.create("local://127.0.0.1:8080"), NodeVersion.UNKNOWN, false);
+    private static final InternalNode NODE_2 = new InternalNode("node-2", URI.create("local://127.0.0.1:8081"), NodeVersion.UNKNOWN, false);
+    private static final InternalNode NODE_3 = new InternalNode("node-3", URI.create("local://127.0.0.1:8082"), NodeVersion.UNKNOWN, false);
+
+    private FinalizerService finalizerService;
+    private NodeTaskMap nodeTaskMap;
+
+    @BeforeClass
+    public void beforeClass()
+    {
+        finalizerService = new FinalizerService();
+        finalizerService.start();
+        nodeTaskMap = new NodeTaskMap(finalizerService);
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void afterClass()
+    {
+        nodeTaskMap = null;
+        if (finalizerService != null) {
+            finalizerService.destroy();
+            finalizerService = null;
+        }
+    }
+
+    @Test
+    public void testHappyPath()
+            throws Exception
+    {
+        TestingRemoteTaskFactory remoteTaskFactory = new TestingRemoteTaskFactory();
+        TestingTaskSourceFactory taskSourceFactory = createTaskSourceFactory(5, 2);
+        TestingNodeSupplier nodeSupplier = TestingNodeSupplier.create(ImmutableMap.of(
+                NODE_1, ImmutableList.of(CATALOG),
+                NODE_2, ImmutableList.of(CATALOG),
+                NODE_3, ImmutableList.of(CATALOG)));
+
+        TestingExchange sinkExchange = new TestingExchange(false);
+
+        TestingExchange sourceExchange1 = new TestingExchange(false);
+        TestingExchange sourceExchange2 = new TestingExchange(false);
+
+        FaultTolerantStageScheduler scheduler = createFaultTolerantTaskScheduler(
+                remoteTaskFactory,
+                taskSourceFactory,
+                createNodeAllocator(nodeSupplier),
+                TaskLifecycleListener.NO_OP,
+                Optional.of(sinkExchange),
+                ImmutableMap.of(SOURCE_FRAGMENT_ID_1, sourceExchange1, SOURCE_FRAGMENT_ID_2, sourceExchange2),
+                2);
+
+        ListenableFuture<Void> blocked = scheduler.isBlocked();
+        assertTrue(blocked.isDone());
+
+        scheduler.schedule();
+
+        blocked = scheduler.isBlocked();
+        // blocked on first source exchange
+        assertFalse(blocked.isDone());
+
+        sourceExchange1.setSourceHandles(ImmutableList.of(new TestingExchangeSourceHandle(0, 1)));
+        // still blocked on the second source exchange
+        assertFalse(blocked.isDone());
```

Comment:
nit add assertBlocked(future) and assertNotBlocked(future) for better readability